File: src/main/java/com/lessmarkup/engine/data/dialects/MicrosoftSqlDialect.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
package com.lessmarkup.engine.data.dialects;

import java.util.OptionalInt;

public class MicrosoftSqlDialect implements DatabaseLanguageDialect {

    @Override
    public String getTypeDeclaration(DatabaseDataType type, OptionalInt sizeLimit, boolean required, String characterSet) {
        String nullable = required ? " NOT NULL" : " NULL";
        switch (type) {
            case INT:
                return "[INT]" + nullable;
            case LONG:
                return "[BIGINT]" + nullable;
            case DATE_TIME:
                return "[DATETIME]" + nullable;
            case IDENTITY:
                return "[BIGINT] IDENTITY(1,1)" + nullable;
            case BOOLEAN:
                return "[BIT]" + nullable;
            case FLOAT:
                return "[FLOAT]" + nullable;
            case DOUBLE:
                return "[DOUBLE]" + nullable;
            case STRING:
                return String.format("[NVARCHAR](%s)%s", sizeLimit.isPresent() ? Integer.toString(sizeLimit.getAsInt()) : "max", nullable);
            case BINARY:
                return String.format("[VARBINARY](%s)%s", sizeLimit.isPresent() ? Integer.toString(sizeLimit.getAsInt()) : "max", nullable);
            default:
                throw new IllegalArgumentException();
        }
    }

    @Override
    public String decorateName(String name) {
        return "[" + name + "]";
    }

    @Override
    public DatabaseDataType getDataType(String dataType) {
        switch (dataType.toUpperCase()) {
            case "INT":
                return DatabaseDataType.INT;
            case "BIGINT":
                return DatabaseDataType.LONG;
            case "DATETIME":
                return DatabaseDataType.DATE_TIME;
            case "BIT":
                return DatabaseDataType.BOOLEAN;
            case "FLOAT":
                return DatabaseDataType.FLOAT;
            case "DOUBLE":
                return DatabaseDataType.DOUBLE;
            case "NVARCHAR":
                return DatabaseDataType.STRING;
            case "VARBINARY":
                return DatabaseDataType.BINARY;
            default:
                return null;
        }
    }

    @Override
    public String paging(int from, int count) {
        return String.format("OFFSET %d ROWS FETCH NEXT %d ROWS ONLY", from, count);
    }
}

File: src/main/scala/com/lessmarkup/engine/data/dialects/MicrosoftSqlDialect.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.engine.data.dialects

class MicrosoftSqlDialect extends DatabaseLanguageDialect {

  def getTypeDeclaration(dataType: DatabaseDataType, sizeLimit: Option[Int], required: Boolean, characterSet: Option[String]): String = {
    val nullable: String = if (required) " NOT NULL" else " NULL"
    dataType match {
      case DatabaseDataType.INT => "[INT]" + nullable
      case DatabaseDataType.LONG => "[BIGINT]" + nullable
      case DatabaseDataType.DATE_TIME => "[DATETIME]" + nullable
      case DatabaseDataType.IDENTITY => "[BIGINT] IDENTITY(1,1)"
      case DatabaseDataType.BOOLEAN => "[BIT]" + nullable
      case DatabaseDataType.FLOAT => "[FLOAT]" + nullable
      case DatabaseDataType.DOUBLE => "[DOUBLE]" + nullable
      case DatabaseDataType.STRING => String.format("[NVARCHAR](%s)%s", if (sizeLimit.isDefined) Integer.toString(sizeLimit.get) else "max", nullable)
      case DatabaseDataType.BINARY => String.format("[VARBINARY](%s)%s", if (sizeLimit.isDefined) Integer.toString(sizeLimit.get) else "max", nullable)
      case _ => throw new IllegalArgumentException
    }
  }

  def getDataType(dataType: String): DatabaseDataType = {
    dataType.toUpperCase match {
      case "INT" => DatabaseDataType.INT
      case "BIGINT" => DatabaseDataType.LONG
      case "DATETIME" => DatabaseDataType.DATE_TIME
      case "BIT" => DatabaseDataType.BOOLEAN
      case "FLOAT" => DatabaseDataType.FLOAT
      case "DOUBLE" => DatabaseDataType.DOUBLE
      case "NVARCHAR" => DatabaseDataType.STRING
      case "VARBINARY" => DatabaseDataType.BINARY
      case _ => null
    }
  }

  def paging(from: Int, count: Int): String = {
    s"OFFSET $from ROWS FETCH NEXT $count ROWS ONLY"
  }

  def decorateName(name: String): String = {
    "[" + name + "]"
  }

  override def schemaTablesName: String = "INFORMATION_SCHEMA.TABLES"

  override def definePrimaryKey: Boolean = true
}
